### ISSUE TITLE
Properties declarations on functions fail compiled when Circular

### ISSUE THREAD

**Issue by bluelovers** [microsoft/TypeScript#28653]
**TypeScript Version:**  3.2.0-dev.201xxxxx

**Search Terms:**

**Code**

```ts
export interface IConstructor<T extends unknown>
{
	new(...argv): T;
	readonly prototype: T;
}

// @ts-ignore
const EMPTY_OBJ_PROTOTYPE = {}.__proto__;
const FUNCTION_PROTOTYPE = Function.prototype;

function isExtendsOf<T extends IConstructor<any>>(classConstructor: T,
	targetConstructor: IConstructor,
): classConstructor is T
{
	let prototype;
	let current = classConstructor;

	if (
		!classConstructor
		|| !classConstructor.prototype
		|| !targetConstructor
		|| !targetConstructor.prototype
	)
	{
		throw new TypeError(`not a constructor object`)
	}

	if (classConstructor === targetConstructor)
	{
		return true;
	}

	do
	{
		prototype = Object.getPrototypeOf(current);

		if (prototype === targetConstructor)
		{
			return true;
		}
		else if (prototype === null)
		{
			return false;
		}

		current = prototype;
	}
	while (prototype && !(
		prototype === FUNCTION_PROTOTYPE
		|| prototype === EMPTY_OBJ_PROTOTYPE
	));

	return false;
}

isExtendsOf.isExtendsOf = isExtendsOf;
isExtendsOf.default = isExtendsOf;

export = isExtendsOf

```

**Expected behavior:**

can compiled xxx.d.ts , xxx.js

**Actual behavior:**

fail compiled xxx.d.ts , xxx.js

**Playground Link:** 

**Related Issues:**

**Comment by bluelovers** [comment]
## base code

```ts
function A()
{
	return 0;
}

A.A = A

export = A
```

=>

Actual behavior:

```ts
declare function A(): number;
declare namespace A {
    var A: typeof A;
}
export = A;
```

## fail compiled and no any error msg

![image](https://user-images.githubusercontent.com/167966/49312485-4e39c400-f51f-11e8-99b8-b3c4478eebe8.png)

```ts
export type IA = 0

function A()
{
	return 0;
}

A.A = A

export = A
```

=>

Expected behavior:

```ts
export declare type IA = 0;
declare function A(): number;
declare namespace A {
    var A: typeof A;
}
export = A;
```

Actual behavior:

```ts
declare function A(): number;
declare namespace A {
    var A: typeof A;
}
export = A;
```

## another code

```ts
export type IA = 0

function A()
{
	return 0;
}

namespace A{}

A.A = A

export = A
```

=>

```ts
export declare type IA = 0;
declare function A(): number;
declare namespace A {
    var A: {
        (): number;
        A: any;
    };
}
declare namespace A { }
export = A;
```

**Comment by RyanCavanaugh** [comment]
Seems to have been fixed by 3.5.3:
```
a.ts:3:6 - error TS7019: Rest parameter 'argv' implicitly has an 'any[]' type.

3  new(...argv): T;
       ~~~~~~~

a.ts:12:21 - error TS2314: Generic type 'IConstructor<T>' requires 1 type argument(s).

12  targetConstructor: IConstructor,
                       ~~~~~~~~~~~~

a.ts:56:1 - error TS4033: Property 'isExtendsOf' of exported interface has or is using private name 'isExtendsOf'.

56 isExtendsOf.isExtendsOf = isExtendsOf;
   ~~~~~~~~~~~~~~~~~~~~~~~

a.ts:57:1 - error TS4033: Property 'default' of exported interface has or is using private name 'isExtendsOf'.

57 isExtendsOf.default = isExtendsOf;
   ~~~~~~~~~~~~~~~~~~~

a.ts:59:1 - error TS2309: An export assignment cannot be used in a module with other exported elements.

59 export = isExtendsOf
   ~~~~~~~~~~~~~~~~~~~~
```
